working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,937 @@
+constructor invocation, which makes `E<T1, .. Tk>(o).m()` ambiguous when
+specifying extensions that may or may not have a constructor. The use of
+`invokeExtensionMethod` makes it explicit and unambiguous that we are
+talking about an extension method invocation.*
+
+The static analysis of `invokeExtensionMethod` is that it takes exactly two
+positional arguments and must be the receiver in a member access. The first
+argument must be a `<type>`, denoting an extension type _T_, and the second
+argument must be an expression whose static type is _T_ or the
+corresponding instantiated on-type. The member access must be a member of
+`E`. If the member access is a method invocation (including an invocation
+of an operator that takes at least one argument), it is allowed to pass an
+actual argument list, and the static analysis of the actual arguments
+proceeds as with other function calls, using a signature where the formal
+type parameters of `E` are replaced by `T1, .. Tk`. The type of the entire
+member access is the return type of said member if it is a member
+invocation, and the function type of the method if it is an extension
+member tear-off, again substituting `T1, .. Tk` for the formal type
+parameters.
+
+Assume that _E_ is an extension declaration of the following form:
+
+```dart
+extension Ext<X1 extends B1, .. Xk extends Bk> on T {
+  ... // Members
+}
+```
+
+*Note that `extension type` declarations and other enhanced forms are
+discussed in later sections. The properties specified here are valid for
+those forms as well, except for the differences which are specified
+explicitly.*
+
+It is then allowed to use `Ext<S1, .. Sk>` as a type.
+
+*For example, it can occur as the declared type of a variable or parameter,
+as the return type of a function or getter, as a type argument in a type,
+as the on-type of an extension, as the type in the `onPart` of a
+try/catch statement, or in a type test `o is E` or a type cast `o as E`, or
+as the body of a type alias. It is also allowed to create a new instance
+where one or more extension types occur as type arguments.*
+
+When `k` is zero, `Ext<S1, .. Sk>` simply stands for `Ext`, a non-generic
+extension. When `k` is greater than zero, a raw occurrence `Ext` is treated
+like a raw type: Instantiation to bound is used to obtain the omitted type
+arguments.
+
+We say that the static type of said variable, parameter, etc. _is the
+extension type_ `Ext<S1, .. Sk>`, and that its static type _is an extension
+type_.
+
+A compile-time error occurs if an extension type is used as a
+superinterface of a class or mixin, or if an extension type is used to
+derive a mixin.
+
+*So `class C extends E1 with E2 implements E3 {}` has three errors if `E1`,
+`E2`, and `E3` are extension types, and `mixin M on E1 implements E2 {}`
+has two errors.*
+
+If `e` is an expression whose static type is the extension type
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>,
```

Comment:
I don't think this is really true if `e` is `this`, and `m` is in the lexical scope.

```
extension Ext1 on int {}
extension Ext2 on Ext1 {
  void test() {
    this.test(); // `this` has static type `Ext1` but this is not treated as `invokeExtensionMethod(Ext1, this).test()`
  }
}
```